File: luckyj/shanten.py
"""Four-player shanten. 0=tenpai, -1=complete; fixed melds inferred from length.

The regular-hand DFS below is adapted from MahjongRepository/mahjong,
mahjong/shanten.py at d16e68f378a7ff4457073f9eec52c5f8d965e03c (MIT).
Copyright (c) 2017 mahjong Python library contributors. See licenses/mahjong.txt.
Changes: four-player-only API, compact state updates, validation, bounded cache.
Special hands are disabled whenever a fixed meld (including ankan) exists.
"""
from functools import lru_cache

ORPHANS = (0, 8, 9, 17, 18, 26, 27, 28, 29, 30, 31, 32, 33)


from .native_accel import load
_NATIVE = load()


@lru_cache(maxsize=200_000)
def shanten(hand: tuple[int, ...]) -> int:
    n = sum(hand)
    if len(hand) != 34 or n not in (1,2,4,5,7,8,10,11,13,14) or any(type(x) is not int or not 0 <= x <= 4 for x in hand):
        raise ValueError('Invalid 34-count hand (1..14 tiles, length 3n+1 or 3n+2)')
    best = regular(hand)
    if n >= 13:
        pairs = sum(x >= 2 for x in hand)
        kinds = sum(x > 0 for x in hand)
        chiitoi = 6 - pairs + max(0, 7-kinds)
        kokushi = 13 - sum(hand[i] > 0 for i in ORPHANS) - int(any(hand[i] >= 2 for i in ORPHANS))
        best = min(best, chiitoi, kokushi)
    return best


class _Regular:
    def __init__(self, tiles):
        self.h = list(tiles)
        self.m = self.t = self.p = self.j = self.four = self.iso = 0
        self.best = 8

    def calculate(self, n):
        four = isolated = 0
        for i in range(27,34):
            count = self.h[i]
            if count == 4:
                self.m += 1; self.j += 1; four |= 1 << (i-27); isolated |= 1 << (i-27)
            elif count == 3: self.m += 1
            elif count == 2: self.p += 1
            elif count == 1: isolated |= 1 << (i-27)
        if self.j and n % 3 == 2: self.j -= 1
        if isolated:
            self.iso |= 1 << 27
            if (four | isolated) == four: self.four |= 1 << 27
        for i in range(27): self.four |= (self.h[i] == 4) << i
        self.m += (14-n)//3
        self.run(0)
        return self.best

    def branch(self, indices, field, depth):
        for i in indices: self.h[i] -= 1
        setattr(self, field, getattr(self, field)+1)
        self.run(depth)
        setattr(self, field, getattr(self, field)-1)
        for i in indices: self.h[i] += 1

    def isolated(self, k):
        self.h[k] -= 1; self.iso |= 1 << k
        self.run(k+1)
        self.h[k] += 1; self.iso &= ~(1 << k)

    def run(self, k):
        if self.best == -1: return
        while k < 27 and not self.h[k]: k += 1
        if k == 27:
            value = 8-2*self.m-self.t-self.p
            candidates = self.m+self.t
            if self.p: candidates += self.p-1
            elif self.four and self.iso and (self.four | self.iso) == self.four: value += 1
            value += max(0,candidates-4)
            if value != -1: value = max(value,self.j)
            self.best = min(self.best,value)
            return
        n = k % 9
        if self.h[k] == 4:
            self.h[k] -= 3; self.m += 1
            if n < 7 and self.h[k+2]:
                if self.h[k+1]: self.branch((k,k+1,k+2),'m',k+1)
                self.branch((k,k+2),'t',k+1)
            if n < 8 and self.h[k+1]: self.branch((k,k+1),'t',k+1)
            self.isolated(k)
            self.h[k] += 3; self.m -= 1
            self.h[k] -= 2; self.p += 1
            if n < 7 and self.h[k+2]:
                if self.h[k+1]: self.branch((k,k+1,k+2),'m',k)
                self.branch((k,k+2),'t',k+1)
            if n < 8 and self.h[k+1]: self.branch((k,k+1),'t',k+1)
            self.h[k] += 2; self.p -= 1
        if self.h[k] == 3:
            self.branch((k,k,k),'m',k+1)
            self.h[k] -= 2; self.p += 1
            if n < 7 and self.h[k+1] and self.h[k+2]: self.branch((k,k+1,k+2),'m',k+1)
            else:
                if n < 7 and self.h[k+2]: self.branch((k,k+2),'t',k+1)
                if n < 8 and self.h[k+1]: self.branch((k,k+1),'t',k+1)
            self.h[k] += 2; self.p -= 1
            if n < 7 and self.h[k+1] >= 2 and self.h[k+2] >= 2:
                for i in (k,k+1,k+2): self.h[i] -= 2
                self.m += 2; self.run(k); self.m -= 2
                for i in (k,k+1,k+2): self.h[i] += 2
        if self.h[k] == 2:
            self.branch((k,k),'p',k+1)
            if n < 7 and self.h[k+1] and self.h[k+2]: self.branch((k,k+1,k+2),'m',k)
        if self.h[k] == 1:
            if n < 6 and self.h[k+1] == 1 and self.h[k+2] and self.h[k+3] != 4:
                self.branch((k,k+1,k+2),'m',k+2)
            else:
                self.isolated(k)
                if n < 7 and self.h[k+2]:
                    if self.h[k+1]: self.branch((k,k+1,k+2),'m',k+1)
                    self.branch((k,k+2),'t',k+1)
                if n < 8 and self.h[k+1]: self.branch((k,k+1),'t',k+1)


def regular(hand):
    return _NATIVE(bytes(hand)) if _NATIVE is not None else _Regular(hand).calculate(sum(hand))

File: luckyj/native/shanten.c
/* Four-player regular shanten, adapted from MahjongRepository/mahjong
 * d16e68f378a7ff4457073f9eec52c5f8d965e03c, MIT. See licenses/mahjong.txt.
 * Copyright (c) 2017 mahjong Python library contributors.
 * Optional accelerator; the Python implementation remains the fallback. */
#include <stdint.h>
typedef struct {int h[34],m,t,p,j,best; uint32_t four,iso;} State;
static void run(State *s,int k);
static void branch(State*s,int a,int b,int c,int *field,int depth) {
    s->h[a]--; if(b>=0)s->h[b]--; if(c>=0)s->h[c]--; (*field)++;
    run(s,depth);
    (*field)--;s->h[a]++; if(b>=0)s->h[b]++; if(c>=0)s->h[c]++;
}
static void isolated(State*s,int k) {
    s->h[k]--;s->iso |= 1u<<k;run(s,k+1);s->h[k]++;s->iso &= ~(1u<<k);
}
static void run(State*s,int k) {
    if(s->best==-1)return;
    while(k<27&&!s->h[k])k++;
    if(k==27) {
        int value=8-2*s->m-s->t-s->p, candidates=s->m+s->t;
        if(s->p)candidates+=s->p-1;
        else if(s->four&&s->iso&&(s->four|s->iso)==s->four)value++;
        if(candidates>4)value+=candidates-4;
        if(value!=-1&&value<s->j)value=s->j;
        if(value<s->best)s->best=value;
        return;
    }
    int n=k%9;
    if(s->h[k]==4) {
        s->h[k]-=3;s->m++;
        if(n<7&&s->h[k+2]) {
            if(s->h[k+1])branch(s,k,k+1,k+2,&s->m,k+1);
            branch(s,k,k+2,-1,&s->t,k+1);
        }
        if(n<8&&s->h[k+1])branch(s,k,k+1,-1,&s->t,k+1);
        isolated(s,k);s->h[k]+=3;s->m--;
        s->h[k]-=2;s->p++;
        if(n<7&&s->h[k+2]) {
            if(s->h[k+1])branch(s,k,k+1,k+2,&s->m,k);
            branch(s,k,k+2,-1,&s->t,k+1);
        }
        if(n<8&&s->h[k+1])branch(s,k,k+1,-1,&s->t,k+1);
        s->h[k]+=2;s->p--;
    }
    if(s->h[k]==3) {
        branch(s,k,k,k,&s->m,k+1);s->h[k]-=2;s->p++;
        if(n<7&&s->h[k+1]&&s->h[k+2])branch(s,k,k+1,k+2,&s->m,k+1);
        else {
            if(n<7&&s->h[k+2])branch(s,k,k+2,-1,&s->t,k+1);
            if(n<8&&s->h[k+1])branch(s,k,k+1,-1,&s->t,k+1);
        }
        s->h[k]+=2;s->p--;
        if(n<7&&s->h[k+1]>=2&&s->h[k+2]>=2) {
            s->h[k]-=2;s->h[k+1]-=2;s->h[k+2]-=2;s->m+=2;
            run(s,k);
            s->m-=2;s->h[k]+=2;s->h[k+1]+=2;s->h[k+2]+=2;
        }
    }
    if(s->h[k]==2) {
        branch(s,k,k,-1,&s->p,k+1);
        if(n<7&&s->h[k+1]&&s->h[k+2])branch(s,k,k+1,k+2,&s->m,k);
    }
    if(s->h[k]==1) {
        if(n<6&&s->h[k+1]==1&&s->h[k+2]&&s->h[k+3]!=4)branch(s,k,k+1,k+2,&s->m,k+2);
        else {
            isolated(s,k);
            if(n<7&&s->h[k+2]) {
                if(s->h[k+1])branch(s,k,k+1,k+2,&s->m,k+1);
                branch(s,k,k+2,-1,&s->t,k+1);
            }
            if(n<8&&s->h[k+1])branch(s,k,k+1,-1,&s->t,k+1);
        }
    }
}
int luckyj_regular(const unsigned char *h) {
    State s={0};s.best=8;int n=0;uint32_t f=0,isol=0;
    for(int i=0;i<34;i++){s.h[i]=h[i];n+=h[i];if(h[i]>4)return -999;}
    if(n>14||n%3==0)return -999;
    for(int i=27;i<34;i++) {
        if(h[i]==4){s.m++;s.j++;f|=1u<<(i-27);isol|=1u<<(i-27);}
        else if(h[i]==3)s.m++;
        else if(h[i]==2)s.p++;
        else if(h[i]==1)isol|=1u<<(i-27);
    }
    if(s.j&&n%3==2)s.j--;
    if(isol){s.iso|=1u<<27;if((f|isol)==f)s.four|=1u<<27;}
    for(int i=0;i<27;i++)if(h[i]==4)s.four|=1u<<i;
    s.m+=(14-n)/3;run(&s,0);return s.best;
}
